### PR TITLE
showing the history of the calculator

### PR DESCRIPTION
**Pull request by yaaHosein**
let us work together

### CODE REVIEW

**Comment by yaaHosein** [comment]
As you know we have to spectate arrays content by ,
How I can achieve this in our current history array

history = [this.history]

**Comment by yaaHosein** [comment]
2)) Please why the calculator buttons stopped working?

**Comment by SMH110** [comment]
> 2)) Please why the calculator buttons stopped working?

Because you have overwritten the `onButtonNumberClick` 

when a button clicked, it calls `this.onButtonNumberClick`  and `this` refers to the new class (`DevelopedCalculator `)

it was working beore adding  `this.onButtonNumberClick`  because the javascript engine try to search for `onButtonNumberClick` on the `DevelopedCalculator ` class and it does not find a method there, then it will move to the parent of the ``DevelopedCalculator ` and it will find a method there.
remember the main implementation of the making button number works done in the parent class. i.e. we update the current result and the result element in the parent class.

so to fix this issue, in the  `DevelopedCalculator `  class and inside `onButtonNumberClick` call the `onButtonNumberClick` of the parent class.

**Comment by SMH110** [comment]
> As you know we have to spectate arrays content by ,
> How I can achieve this in our current history array
> 
> history = [this.history]

You're defining the history array twice, I don't know which one you want to use the first or the last?

**Comment by yaaHosein** [comment]
> > 2)) Please why the calculator buttons stopped working?
> 
> Because you have overwritten the `onButtonNumberClick`
> 
> when a button clicked, it calls `this.onButtonNumberClick` and `this` refers to the new class (`DevelopedCalculator `)
> 
> it was working beore adding `this.onButtonNumberClick` because the javascript engine try to search for `onButtonNumberClick` on the `DevelopedCalculator ` class and it does not find a method there, then it will move to the parent of the ``DevelopedCalculator ` and it will find a method there.
> remember the main implementation of the making button number works done in the parent class. i.e. we update the current result and the result element in the parent class.
> 
> so to fix this issue, in the `DevelopedCalculator ` class and inside `onButtonNumberClick` call the `onButtonNumberClick` of the parent class.

Do you refer to before adding 
this.onButtonNumberClick
In the new class or old one 
I am asking because I did not add yet this. before onButtonNumberClick 

Secondly 
Regarding to overwriting I choose the mentioned method ended by _
But you wanted to change it to be same as the one which is in the old class 

Thanks

**Comment by yaaHosein** [comment]
> > As you know we have to spectate arrays content by ,
> > How I can achieve this in our current history array
> > history = [this.history]
> 
> You're defining the history array twice, I don't know which one you want to use the first or the last?

I have to wait until solve the method problem 
Then check the second one has to be in its place or to be moved up

**Comment by SMH110** [comment]
> 
> Do you refer to before adding
> this.onButtonNumberClick
> In the new class or old one
> I am asking because I did not add yet this. before onButtonNumberClick

If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator `

**Comment by yaaHosein** [comment]
> > Do you refer to before adding
> > this.onButtonNumberClick
> > In the new class or old one
> > I am asking because I did not add yet this. before onButtonNumberClick
> 
> If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`

Thanks a lot my dear teacher 
I know I added onButtonNumberClick in the new class 
I am asking about 
this.onButtonNumberClick

I did not added it yet in the new class 

So do you telling me now about that one which is in the new class or old one

**Comment by SMH110** [comment]
> > > Do you refer to before adding
> > > this.onButtonNumberClick
> > > In the new class or old one
> > > I am asking because I did not add yet this. before onButtonNumberClick
> > 
> > 
> > If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`
> 
> Thanks a lot my dear teacher
> I know I added onButtonNumberClick in the new class
> I am asking about
> this.onButtonNumberClick

`this.onButtonNumberClick` = `onButtonNumberClick`

**Comment by yaaHosein** [comment]
> > > > Do you refer to before adding
> > > > this.onButtonNumberClick
> > > > In the new class or old one
> > > > I am asking because I did not add yet this. before onButtonNumberClick
> > > 
> > > 
> > > If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`
> > 
> > 
> > Thanks a lot my dear teacher
> > I know I added onButtonNumberClick in the new class
> > I am asking about
> > this.onButtonNumberClick
> 
> `this.onButtonNumberClick` = `onButtonNumberClick`

Ok thanks for clearing this 
Please how I can call onButtonNumberClick method inside the child one class same method 

I tried 
this.onButtonNumberClick()

But it was a big error in the console panel

**Comment by SMH110** [comment]
> > > > > Do you refer to before adding
> > > > > this.onButtonNumberClick
> > > > > In the new class or old one
> > > > > I am asking because I did not add yet this. before onButtonNumberClick
> > > > 
> > > > 
> > > > If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`
> > > 
> > > 
> > > Thanks a lot my dear teacher
> > > I know I added onButtonNumberClick in the new class
> > > I am asking about
> > > this.onButtonNumberClick
> > 
> > 
> > `this.onButtonNumberClick` = `onButtonNumberClick`
> 
> Ok thanks for clearing this
> Please how I can call onButtonNumberClick method inside the child one class same method
> 
> I tried
> this.onButtonNumberClick()
> 
> But it was a big error in the console panel

How did you call the render method of the parent class?

**Comment by yaaHosein** [comment]
> > > > > > Do you refer to before adding
> > > > > > this.onButtonNumberClick
> > > > > > In the new class or old one
> > > > > > I am asking because I did not add yet this. before onButtonNumberClick
> > > > > 
> > > > > 
> > > > > If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`
> > > > 
> > > > 
> > > > Thanks a lot my dear teacher
> > > > I know I added onButtonNumberClick in the new class
> > > > I am asking about
> > > > this.onButtonNumberClick
> > > 
> > > 
> > > `this.onButtonNumberClick` = `onButtonNumberClick`
> > 
> > 
> > Ok thanks for clearing this
> > Please how I can call onButtonNumberClick method inside the child one class same method
> > I tried
> > this.onButtonNumberClick()
> > But it was a big error in the console panel
> 
> How did you call the render method of the parent class?

If you remember you asked me to remove the instant class in the parent class 
If  calling render in the child’s class is not calling the parent’s one also 
I am not calling the parent’s one 
And do know know how to call it again 
Thanks

**Comment by SMH110** [comment]
> > > > > > > Do you refer to before adding
> > > > > > > this.onButtonNumberClick
> > > > > > > In the new class or old one
> > > > > > > I am asking because I did not add yet this. before onButtonNumberClick
> > > > > > 
> > > > > > 
> > > > > > If you look carefully you will see that you added `onButtonNumberClick` to `DevelopedCalculator`
> > > > > 
> > > > > 
> > > > > Thanks a lot my dear teacher
> > > > > I know I added onButtonNumberClick in the new class
> > > > > I am asking about
> > > > > this.onButtonNumberClick
> > > > 
> > > > 
> > > > `this.onButtonNumberClick` = `onButtonNumberClick`
> > > 
> > > 
> > > Ok thanks for clearing this
> > > Please how I can call onButtonNumberClick method inside the child one class same method
> > > I tried
> > > this.onButtonNumberClick()
> > > But it was a big error in the console panel
> > 
> > 
> > How did you call the render method of the parent class?
> 
> If you remember you asked me to remove the instant class in the parent class
> If calling render in the child’s class is not calling the parent’s one also
> I am not calling the parent’s one
> And do know know how to call it again
> Thanks

Can you explain line by line your render method

**Comment by yaaHosein** [comment]
In the parent class or in the child one  !?

**Comment by SMH110** [comment]
> In the parent class or in the child one !?

You are now working on DevelopedCalculator
So all my questions are related to this class unless i say

**Comment by yaaHosein** [comment]
4- Calling the render method of parent 
6- Creating div
8- appending a string and the current result to the above div
10- prepending the history div in the host element

**Comment by SMH110** [comment]
> 4- Calling the render method of parent

in the render method you're calling the render method of the parent class.
I asked you to explain the render method to refresh you memory and remind you of how to can access the parent class.
now, in the `onButtonNumberClick` call the `onButtonNumberClick` of the parent class.

**Comment by yaaHosein** [comment]
> > 4- Calling the render method of parent
> 
> in the render method you're calling the render method of the parent class.
> I asked you to explain the render method to refresh you memory and remind you of how to can access the parent class.
> now, in the `onButtonNumberClick` call the `onButtonNumberClick` of the parent class.

**Comment by SMH110** [comment]
> 4- Calling the render method of parent
> 
> in the render method you're calling the render method of the parent class.
> I asked you to explain the render method to refresh you memory and remind you of how to can access the parent class.
> now, in the onButtonNumberClick call the onButtonNumberClick of the parent class.

Image is not opening

**Comment by yaaHosein** [comment]
Thanks 
Yes I know 
I have sent it to your email

**Comment by yaaHosein** [comment]
In the image
I am showing you 
Also when I added 
super.onButtonNumberClick()
Inside this method I got error message 
Can not read property target of undefined (( line 134))

Also I tried 

super.this.onButtonNumberClick()

I got error message

**Comment by SMH110** [comment]
> In the image
> I am showing you
> Also when I added
> super.onButtonNumberClick()
> Inside this method I got error message
> Can not read property target of undefined (( line 134))
> 
> Also I tried
> 
> super.this.onButtonNumberClick()
> 
> I got error message

Push your changes pls

**Comment by SMH110** [comment]
Looking at your changes, I can see you were very close to fix this issue.

Previously, when you created the `Calculator` class; when a button clicked the `onButtonNumberClick` used to be called and click event object was passed to that method. 

when you created the new class which inherits from `Calculator` class and before adding  `onButtonNumberClick` on the new class. when a button clicked the JS engine called  `this.onButtonNumberClick` and because the method wasn't defined on the new class, the engine searched for a method with same name on the parent class and found it.

when you added the method to the new class, the method is being called every time a button clicked.
the click event object is now get passed to the  `onButtonNumberClick`on the new class.
you're calling the  `onButtonNumberClick` of the parent inside  `onButtonNumberClick` so you need to pass the event object to the  `onButtonNumberClick` of the parent class 

to that:
1. make the  `onButtonNumberClick` take an event paramter 
2. when you call  `super.onButtonNumberClick()` pass the event to method

**Comment by yaaHosein** [comment]
Please why I can not see the history value yet?!
Thanks

**Comment by SMH110** [comment]
next, we need to update the history element

First: you need to have access to the history element which you created in the `render` method.
in other words, you need to select/query the history element and update its value every time a button clicked.
I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.

**Comment by yaaHosein** [comment]
> next, we need to update the history element
> 
> First: you need to have access to the history element which you created in the `render` method.
> in other words, you need to select/query the history element and update its value every time a button clicked.
> I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.

As you know we are updating the result by onButtonCNumberlick method 
So please which new update you are mentioning?!

**Comment by yaaHosein** [comment]
> > next, we need to update the history element
> > First: you need to have access to the history element which you created in the `render` method.
> > in other words, you need to select/query the history element and update its value every time a button clicked.
> > I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.
> 
> As you know we are updating the result by onButtonCNumberlick method
> So please which new update you are mentioning?!

Please why we don’t have access to history element?!

**Comment by SMH110** [comment]
> > next, we need to update the history element
> > First: you need to have access to the history element which you created in the `render` method.
> > in other words, you need to select/query the history element and update its value every time a button clicked.
> > I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.
> 
> As you know we are updating the result by onButtonCNumberlick method
> So please which new update you are mentioning?!

The history element, 
the history element will show the history of the clicked buttons 

if I click on these buttons in order 1 + 2 = 
I will see 3 in the result element, but I want to see  1, 2  in the history element

**Comment by SMH110** [comment]
> > > next, we need to update the history element
> > > First: you need to have access to the history element which you created in the `render` method.
> > > in other words, you need to select/query the history element and update its value every time a button clicked.
> > > I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.
> > 
> > 
> > As you know we are updating the result by onButtonCNumberlick method
> > So please which new update you are mentioning?!
> 
> Please why we don’t have access to history element?!

to update its value

File: DevelopedCalculator.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 class DevelopedCalculator extends Calculator {
     history = [this.currentResult];
+    historyDiv = null;
+    historyDiv = document.querySelector('div');
```
Comment (SMH110) [inline DevelopedCalculator.js]: `document.querySelector` returns the first match element. in this case it searching in the DOM and returning the first `div` element.

are you sure the history div is the first div element in the page and are you sure that it will be always the first div element in page?

also, I though you're creating the history element in the `render` method

**Comment by SMH110** [comment]
you need to save the history div as property on the class,

first, you need to remember where is this div coming from? 
where did you create it?
was this element in the page before creating the calculator or it's created during creating the calculator?

second, you need to remember how to save something as property on a class 

when you remember the above points you should be able to set the history div/element as property on the class

this step is prerequisite to the next task I will ask you to do

**Comment by yaaHosein** [comment]
> > > next, we need to update the history element
> > > First: you need to have access to the history element which you created in the `render` method.
> > > in other words, you need to select/query the history element and update its value every time a button clicked.
> > > I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.
> > 
> > 
> > As you know we are updating the result by onButtonCNumberlick method
> > So please which new update you are mentioning?!
> 
> The history element,
> the history element will show the history of the clicked buttons
> 
> if I click on these buttons in order 1 + 2 =
> I will see 3 in the result element, but I want to see 1, 2 in the history element

Ok Thanks

File: DevelopedCalculator.js (hunk 1 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: When I create it inside the method i get error message

**Comment by yaaHosein** [comment]
I have pushed some updates 
I wanted to check regarding to more than one time 
Do you were referring to change the query to Allquery
Or to use 
forEach method?!

Thanks

**Comment by SMH110** [comment]
> I have pushed some updates
> I wanted to check regarding to more than one time
> Do you were referring to change the query to Allquery
> Or to use
> forEach method?!
> 
> Thanks

you don't need forEach now, let's focus on saving the history element as property on the class first.

**Comment by SMH110** [comment]
I looked at your changes and now I see you have 
`const historyDiv` and `this.historyDiv` and `this.historyElement` 
can you tell me more about them?
what is the role of each one of them and what is the different between ``this.historyDiv` and `this.historyElement`

File: DevelopedCalculator.js (hunk 1 above)
Comment (SMH110) [reply DevelopedCalculator.js]: > When I create it inside the method i get error message

creating  what?

**Comment by yaaHosein** [comment]
> I looked at your changes and now I see you have
> `const historyDiv` and `this.historyDiv` and `this.historyElement`
> can you tell me more about them?
> what is the role of each one of them and what is the different between ``this.historyDiv`and`this.historyElement`

Unfortunately I don’t know now 
Just I am trying to do what you are asking me step by step 
I hope when I finish and the picture completed understand what is happening

**Comment by yaaHosein** [comment]
Just I am trying to do something which I did similar 

historyElement instead of resultElement
historyDiv instead of resultDiv

**Comment by yaaHosein** [comment]
> > I have pushed some updates
> > I wanted to check regarding to more than one time
> > Do you were referring to change the query to Allquery
> > Or to use
> > forEach method?!
> > Thanks
> 
> you don't need forEach now, let's focus on saving the history element as property on the class first.

Only we have one div
I do not know why querySelector was not correct ??

If you check this tutorial you see there is more than one div 

So the teacher is using querySelectorAll

querySelectorAll method (geekyshows)

**Comment by yaaHosein** [comment]
> > > > next, we need to update the history element
> > > > First: you need to have access to the history element which you created in the `render` method.
> > > > in other words, you need to select/query the history element and update its value every time a button clicked.
> > > > I would suggest saving the history element as property on the class, we did something similar before and you should know how to do that by now.
> > > 
> > > 
> > > As you know we are updating the result by onButtonCNumberlick method
> > > So please which new update you are mentioning?!
> > 
> > 
> > Please why we don’t have access to history element?!
> 
> to update its value

If I was not had access to history element (historyDiv) then how I used them with appendChild and in line 10

**Comment by SMH110** [comment]
> > > I have pushed some updates
> > > I wanted to check regarding to more than one time
> > > Do you were referring to change the query to Allquery
> > > Or to use
> > > forEach method?!
> > > Thanks
> > 
> > 
> > you don't need forEach now, let's focus on saving the history element as property on the class first.
> 
> Only we have one div
> I do not know why querySelector was not correct ??
> 
> If you check this tutorial you see there is more than one div
> 
> So the teacher is using querySelectorAll
> 
> querySelectorAll method (geekyshows)


`querySelectorAll ` returns a list of elements and in your code it returning all `div` elements.
I don't know how you're going to find the `div` that represent the history div ?? 


can you please explain what you did here?

`  this.historyDiv = document.createElement("div");`

**Comment by yaaHosein** [comment]
If you watch the tutorial which I sent you about querySelectorAll you will find there are many divs 
And he is using the mentioned selector to select them 

Now you are telling me this method is returning a list of div
What kind of div
I mean I have one div for seeing the history 
So what more divs we need ?!

Do you mean it will create the new many div
I mean they will not be ready from before to be selected as in the tutorial?!

2))
I am saving the history element as a property on the class 
And for doing that I needed to change const historyDiv to this.history 
already  I titled the pushed related comment to explain what I did

**Comment by yaaHosein** [comment]
> > > > I have pushed some updates
> > > > I wanted to check regarding to more than one time
> > > > Do you were referring to change the query to Allquery
> > > > Or to use
> > > > forEach method?!
> > > > Thanks
> > > 
> > > 
> > > you don't need forEach now, let's focus on saving the history element as property on the class first.
> > 
> > 
> > Only we have one div
> > I do not know why querySelector was not correct ??
> > If you check this tutorial you see there is more than one div
> > So the teacher is using querySelectorAll
> > querySelectorAll method (geekyshows)
> 
> `querySelectorAll ` returns a list of elements and in your code it returning all `div` elements.
> I don't know how you're going to find the `div` that represent the history div ??
> 
> can you please explain what you did here?
> 
> ` this.historyDiv = document.createElement("div");`
If you watch the tutorial which I sent you about querySelectorAll you will find there are many divs 
And he is using the mentioned selector to select them 

Now you are telling me this method is returning a list of div
What kind of div
I mean I have one div for seeing the history 
So what more divs we need ?!

Do you mean it will create the new many div
I mean they will not be ready from before to be selected as in the tutorial?!

2))
I am saving the history element as a property on the class 
And for doing that I needed to change const historyDiv to this.history 
already  I titled the pushed related comment to explain what I did

**Comment by SMH110** [comment]
> Now you are telling me this method is returning a list of div
> What kind of div
> I mean I have one div for seeing the history
> So what more divs we need ?!

` querySelectorAll` is method that can be used to select all elements with a selector 
you have this `document. querySelectorAll("div")`  which returns all the div elements in the page.
I was asking, how can you tell which div is the history div, what if page has 100000000 div elements.

> 
> Do you mean it will create the new many div
> I mean they will not be ready from before to be selected as in the tutorial?!
> 

No, it does not create anything 

> 2))
> I am saving the history element as a property on the class
> And for doing that I needed to change const historyDiv to this.history
> already I titled the pushed related comment to explain what I did


you said, you're  saving the history element as a property on the class
I know what you're doing and I asked you because I wanted you to repeat what you did.
now that you give me what I wanted I have another question 

isn't this the main requirement I asked for?
If yes, then why you're having this code
`  this.historyDiv = document.querySelectorAll('div');`

I mean think of it, you said that this line is the line which is responsible of saving the history element as property on the class
`        this.historyDiv = document.createElement("div");`

and this what I asked you to do long time ago.
so why do you have this line in the render method?
`        this.historyDiv = document.querySelectorAll('div');`
if you are the one who creating the div in javascript, then why you're using `querySelectorAll`?
if you have acess to the div element because you have created it, why confusing yourself with other thing?

**Comment by yaaHosein** [comment]
> > Now you are telling me this method is returning a list of div
> > What kind of div
> > I mean I have one div for seeing the history
> > So what more divs we need ?!
> 
> ` querySelectorAll` is method that can be used to select all elements with a selector
> you have this `document. querySelectorAll("div")` which returns all the div elements in the page.
> I was asking, how can you tell which div is the history div, what if page has 100000000 div elements.
> 
> > Do you mean it will create the new many div
> > I mean they will not be ready from before to be selected as in the tutorial?!
> 
> No, it does not create anything
> 
> > 2))
> > I am saving the history element as a property on the class
> > And for doing that I needed to change const historyDiv to this.history
> > already I titled the pushed related comment to explain what I did
> 
> you said, you're saving the history element as a property on the class
> I know what you're doing and I asked you because I wanted you to repeat what you did.
> now that you give me what I wanted I have another question
> 
> isn't this the main requirement I asked for?
> If yes, then why you're having this code
> ` this.historyDiv = document.querySelectorAll('div');`
> 
> I mean think of it, you said that this line is the line which is responsible of saving the history element as property on the class
> ` this.historyDiv = document.createElement("div");`
> 
> and this what I asked you to do long time ago.
> so why do you have this line in the render method?
> ` this.historyDiv = document.querySelectorAll('div');`
> if you are the one who creating the div in javascript, then why you're using `querySelectorAll`?
> if you have acess to the div element because you have created it, why confusing yourself with other thing?

Thanks a lot
If I have a lot of div then I will get the help from their id or class names 

.....
I am using query selector because of the following:

First: you need to have access to the history element which you created in the render method.
in other words, you need to select/query the history element and update its value every time a button clicked.

Two red lines under query 

Ok I will remove the query selector line when I will be near the laptop

**Comment by SMH110** [comment]
> Thanks a lot
> If I have a lot of div then I will get the help from their id or class names
> 
> .....
> I am using query selector because of the following:
> 
> First: you need to have access to the history element which you created in the render method.
> in other words, you need to select/query the history element and update its value every time a button clicked.
> 
> Two red lines under query
> 
> Ok I will remove the query selector line when I will be near the laptop

that's why I said select/query 
two red line under select 

aso I said "need to have access to the history element " and then tried to explain in different way and said "select/query" 
but I don't know why you chose to look at "query" only

**Comment by yaaHosein** [comment]
> > Thanks a lot
> > If I have a lot of div then I will get the help from their id or class names
> > .....
> > I am using query selector because of the following:
> > First: you need to have access to the history element which you created in the render method.
> > in other words, you need to select/query the history element and update its value every time a button clicked.
> > Two red lines under query
> > Ok I will remove the query selector line when I will be near the laptop
> 
> that's why I said select/query
> two red line under select
> 
> also I said "need to have access to the history element " and then tried to explain in different way and said "select/query"
> but I don't know why you chose to look at "query" only

yes you are right 
I thought I have to add both 

I was thinking before what I have to do with the const before historyDiv and in the same time you want from me to save it as a property 

I was thinking I have to do something else and not touching the const ...

**Comment by yaaHosein** [comment]
1) During creating the calculator

**Comment by SMH110** [comment]
Please add comments to explain your code line by line. unless if you're very confident that you don't need those comments and you can understand the code in the future.

when I say you need to select an elements that does not mean you must need to use one of these:
- document.querySelector 
- document.getElementById
- document.getElementsByClassName 
- ....etc

selecting elements is a generic term, it means getting or accessing a DOM element/node in javascript 
sometimes you create a DOM element/node by writing HTML code and sometimes you create it using javascript 
and based on the way you created the DOM element you query it accordingly.

**Comment by SMH110** [comment]
After addressing the comment above, Please cose/merge this PR into master and create another PR 
you can use the same branch and only need to create a new PR.

in the new PR, add a new method that:
- gets the history array 
- convert the content of the history array to comma separated content. meaning if the array like this [1, 2, 5] the content must be "1,2,5". this is called comma separated.
 if I say  space separated then it should be like this "1 2 5" 
for that you need to research  Array.prototype.join, 
in other words, you need to use the `join` method which's available on every array in javascript 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join

- set the history content which is separated by comma as content of the history div 
for this you need to access the history div and update its content. 

Please put your changes in a new  PR

**Comment by yaaHosein** [comment]
> Please add comments to explain your code line by line. unless if you're very confident that you don't need those comments and you can understand the code in the future.
> 
> when I say you need to select an elements that does not mean you must need to use one of these:
> 
> * document.querySelector
> * document.getElementById
> * document.getElementsByClassName
> * ....etc
> 
> selecting elements is a generic term, it means getting or accessing a DOM element/node in javascript
> sometimes you create a DOM element/node by writing HTML code and sometimes you create it using javascript
> and based on the way you created the DOM element you query it accordingly.

I hope one day soon understand this comment better 
thanks

**Comment by SMH110** [comment]
> > Please add comments to explain your code line by line. unless if you're very confident that you don't need those comments and you can understand the code in the future.
> > when I say you need to select an elements that does not mean you must need to use one of these:
> > 
> > * document.querySelector
> > * document.getElementById
> > * document.getElementsByClassName
> > * ....etc
> > 
> > selecting elements is a generic term, it means getting or accessing a DOM element/node in javascript
> > sometimes you create a DOM element/node by writing HTML code and sometimes you create it using javascript
> > and based on the way you created the DOM element you query it accordingly.
> 
> I hope one day soon understand this comment better
> thanks

Which part you didn't understand?

**Comment by yaaHosein** [comment]
> > > Please add comments to explain your code line by line. unless if you're very confident that you don't need those comments and you can understand the code in the future.
> > > when I say you need to select an elements that does not mean you must need to use one of these:
> > > 
> > > * document.querySelector
> > > * document.getElementById
> > > * document.getElementsByClassName
> > > * ....etc
> > > 
> > > selecting elements is a generic term, it means getting or accessing a DOM element/node in javascript
> > > sometimes you create a DOM element/node by writing HTML code and sometimes you create it using javascript
> > > and based on the way you created the DOM element you query it accordingly.
> > 
> > 
> > I hope one day soon understand this comment better
> > thanks
> 
> Which part you didn't understand?

Difference between DOM and Html and Node

2)) in our project 
Is this will select all div elements?!

this.historyDiv= document.createElement (“div”);

?!?

**Comment by SMH110** [comment]
> Difference between DOM and Html and Node

DOM stands for "Document Object Model" which is the HTML page.
The document has some objects like tree structure.

each HTML element can be called node. 
If you think of HTML page as tree, then each element in that tree is a node 
some nodes contain another nodes and so on.





> 
> 2)) in our project
> Is this will select all div elements?!
> 
> this.historyDiv= document.createElement (“div”);
> 
> ?!?

I don't know why you're mixing between `document.createElement` and `document.querySelectorAll` 
`createElement` creates an element for you 
so instead of typing for example `<div></div>` the `createElement` does that for you.

then it's your responsibility to place that element in the page or in the DOM

while, `querySelectorAll` can be used to select elements by a selector

**Comment by yaaHosein** [comment]
Thanks a lot 
2))
Here :
this.historyDiv= document.createElement (“div”);

I am asking about this keyword (property)

**Comment by yaaHosein** [comment]
Another question please 
I can see only we have one history div
So why  we are using this keyword before it ?!
The content of this div will updated with current result so it makes sense to use this. before history 

But it doesn’t make sense to use this before divHistory 

Because the div is not changing 

Right?!

?!

**Comment by SMH110** [comment]
> Thanks a lot
> 2))
> Here :
> this.historyDiv= document.createElement (“div”);
> 
> I am asking about this keyword (property)

`this` refers to the instance object of the class. 

so if you create two instance  a and b 
in the a instance `this` refers to a and in b instance `this` refers to `b` 

instance = a new copy similar to the original 

so the div which saved in a is not the same div created for b

**Comment by SMH110** [comment]
> Another question please
> I can see only we have one history div
> So why we are using this keyword before it ?!
> The content of this div will updated with current result so it makes sense to use this. before history
> 
> But it doesn’t make sense to use this before divHistory
> 
> Because the div is not changing
> 
> Right?!
> 
> ?!

I didn't understand your question.
your question is missing code examples and code lines

**Comment by yaaHosein** [comment]
> > Thanks a lot
> > 2))
> > Here :
> > this.historyDiv= document.createElement (“div”);
> > I am asking about this keyword (property)
> 
> `this` refers to the instance object of the class.
> 
> so if you create two instance a and b
> in the a instance `this` refers to a and in b instance `this` refers to `b`
> 
> instance = a new copy similar to the original
> 
> so the div which saved in a is not the same div created for b

Ok thanks for reminding me
I forgot this is class calculator 
And I was thinking there is only one calculator and one history div inside it

**Comment by yaaHosein** [comment]
> > Another question please
> > I can see only we have one history div
> > So why we are using this keyword before it ?!
> > The content of this div will updated with current result so it makes sense to use this. before history
> > But it doesn’t make sense to use this before divHistory
> > Because the div is not changing
> > Right?!
> > ?!
> 
> I didn't understand your question.
> your question is missing code examples and code lines

Already you answered this question when you reminded me this is class calculator and we do not have one one history div but maybe 100000 
Thanks
